### PR TITLE
Integrate all parts of jwt authenticator

### PR DESCRIPTION
**Pull request by tzheleznyak**
Integrate parts of jwt authenticator

### CODE REVIEW

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 1]
```diff
@@ -7,7 +7,11 @@ module AuthnJwt
     Authenticate = CommandClass.new(
       dependencies: {
         token_factory: TokenFactory.new,
-        logger: Rails.logger
+        logger: Rails.logger,
+        audit_log: ::Audit.logger,
+        validate_origin: ::Authentication::ValidateOrigin.new,
+        role_cls: ::Role,
+        webservice_class: ::Authentication::Webservice,
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/authenticator.rb:14]: choose between "cls" and "class" and consist with it

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 2]
```diff
@@ -38,11 +47,65 @@ def get_jwt_identity
         @authentication_parameters.jwt_identity =  @jwt_configuration.jwt_identity(@authentication_parameters)
       end
 
+      def validate_origin
+        @validate_origin.(
+          account: @authentication_parameters.account,
+          username: @authentication_parameters.username,
+          client_ip: @authentication_parameters.client_ip
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/authenticator.rb:54]: why dont you use delegators?

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 3]
```diff
@@ -38,11 +47,65 @@ def get_jwt_identity
...
+
       def validate_restrictions
         @logger.debug(LogMessages::Authentication::AuthnJwt::CALLING_VALIDATE_RESTRICTIONS.new)
         @jwt_configuration.validate_restrictions(@authentication_parameters)
       end
 
+      def audit_success
+        @audit_log.log(
+          ::Audit::Event::Authn::Authenticate.new(
+            authenticator_name: @authentication_parameters.authenticator_name,
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/authenticator.rb:66]: use delegator here too

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 4]
```diff
@@ -38,11 +47,65 @@ def get_jwt_identity
...
+            success: true,
+            error_message: nil
+          )
+        )
+      end
+
+      def audit_failure(err)
+        @audit_log.log(
+          ::Audit::Event::Authn::Authenticate.new(
+            authenticator_name: @authentication_parameters.authenticator_name,
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/authenticator.rb:79]: What if the failure is in very first `create_authentication_parameters` stage.
In this case `@authentication_parameters` remains unassigned...

Comment (sashaCher) [reply app/domain/authentication/authn_jwt/authenticator.rb:82]: I guess that `authenticator_input` is enough in order to write the audit

File: app/domain/authentication/authn_jwt/authenticator.rb (hunk 1 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/authenticator.rb:14]: Thanks. Choosed class

File: app/domain/authentication/authn_jwt/authenticator.rb (hunk 2 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/authenticator.rb:57]: I copied it from authenticate.rb
https://github.com/cyberark/conjur/blob/08428cdc8ab192a69f7c01962655d06ab2b62742/app/domain/authentication/authenticate.rb#L67

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 5]
```diff
@@ -41,11 +50,65 @@ def get_jwt_identity
...
+            service: webservice,
+            role_id: audit_role_id,
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/authenticator.rb:84]: Another flow bug is there.
Role will be knows only after jwt signature validation until this moment it's undefined.

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 6]
```diff
@@ -41,16 +54,88 @@ def get_jwt_identity
...
+            account: account,
+            user_id: @authentication_parameters.jwt_identity,
+            privilege: 'authenticate'
+        )
+      end
+
+      def validate_origin
+        @validate_origin.(
+          account: account,
+          username: username,
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/authenticator.rb:69]: I'm not sure that `@authenticator_input.username` has value.
In our case as far as I understand `username` is `@authentication_parameters.jwt_identity`.

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 7]
```diff
@@ -41,16 +54,88 @@ def get_jwt_identity
...
+        )
+      end
+
       def new_token
         @logger.debug(LogMessages::Authentication::AuthnJwt::JWT_AUTHENTICATION_PASSED.new)
         @token_factory.signed_token(
-          account: @authenticator_input.account,
-          username: @authenticator_input.username
+          account: account,
+          username: username
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/authenticator.rb:138]: I'm not sure that `@authenticator_input.username` has value.
In our case as far as I understand username is `@authentication_parameters.jwt_identity`.

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 8]
```diff
@@ -41,16 +54,88 @@ def get_jwt_identity
...
+          IDENTITY_NOT_RETRIEVED_YET,
+          account: account
+        )
+      end
+
+      def audit_role_id
+        ::Audit::Event::Authn::RoleId.new(
+          role: role,
+          account: account,
+          username: username
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/authenticator.rb:122]: username!

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 9]
```diff
@@ -41,16 +54,88 @@ def get_jwt_identity
         @authentication_parameters.jwt_identity =  @jwt_configuration.jwt_identity(@authentication_parameters)
       end
 
+      def validate_user_has_access_to_webservice
+        @validate_role_can_access_webservice.(
+          webservice: webservice,
+            account: account,
+            user_id: @authentication_parameters.jwt_identity,
+            privilege: 'authenticate'
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/authenticator.rb:62]: Do we have it as a const somewhere?

File: app/domain/authentication/authn_jwt/authenticator.rb [hunk 10]
```diff
@@ -41,16 +54,88 @@ def get_jwt_identity
...
+          IDENTITY_NOT_RETRIEVED_YET,
+          account: account
+        )
+      end
+
+      def audit_role_id
+        ::Audit::Event::Authn::RoleId.new(
+          role: role,
+          account: account,
+          username: @authentication_parameters.jwt_identity
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/authenticator.rb:122]: Looks a bit strange. `role` already has username in it...

File: app/domain/authentication/authn_jwt/identity_from_decoded_token_provider.rb [hunk 11]
```diff
@@ -13,7 +13,7 @@ def initialize(authentication_parameters)
       def provide_jwt_identity
         token_field_name = fetch_token_field_name
         @logger.debug(LogMessages::Authentication::AuthnJwt::LOOKING_FOR_IDENTITY_FIELD_NAME.new)
-        jwt_identity = @decoded_token[token_field_name.to_sym]
+        jwt_identity = @decoded_token[token_field_name]
```
Comment (sashaCher) [inline app/domain/authentication/authn_jwt/identity_from_decoded_token_provider.rb:16]: Do we have UT that was able to figure it out?

Comment (sashaCher) [reply app/domain/authentication/authn_jwt/identity_from_decoded_token_provider.rb:16]: Or is it a spec file below :-)

Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/identity_from_decoded_token_provider.rb:16]: Yes

Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/identity_from_decoded_token_provider.rb:16]: https://github.com/cyberark/conjur/blob/feature-jwt-authenticator/spec/app/domain/authentication/authn-jwt/identity_from_decoded_token_provider_spec.rb

File: app/domain/authentication/authn_jwt/authenticator.rb (hunk 9 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/authenticator.rb:62]: I copied pasta it from here 
https://github.com/cyberark/conjur/blob/6f45b33c55d6efcef791d71155d09592a9e1f5c2/app/domain/authentication/authenticate.rb#L63

Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/authenticator.rb:62]: Moved to const only in authn-jwt code

File: app/domain/authentication/authn_jwt/authenticator.rb (hunk 10 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/authenticator.rb:122]: This what happens also here
https://github.com/cyberark/conjur/blob/f0e9a252b13a0a6af95d50d8700d0d3dfc947f02/app/domain/authentication/authenticate.rb#L102

File: Jenkinsfile [hunk 12]
```diff
@@ -557,6 +557,12 @@ pipeline {
         archiveArtifacts artifacts: '*.rpm', fingerprint: true
       }
     }
+
```
Comment (sashaCher) [inline Jenkinsfile:560]: Why are you adding it?

**Review by sashaCher — APPROVED** [review]
LGTM

**Pull request merged**